keps/sig-auth/1688-dynamic-authentication-config/README.md

```diff
@@ -0,0 +1,620 @@
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+This change aims to add a new API Kubernetes REST API called
+`AuthenticationConfig`.  It is similar to the `ValidatingWebhookConfiguration`
+API in that it will be watched by the API server and used to construct a struct.
+In this case, this struct with be an `authenticator.Request`, i.e. a dynamic
+authenticator.  This authenticator will be unioned with the other authenticators.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wishes to use Dex on an existing cluster.  They follow Dex's instructions
+to deploy the IDP onto the cluster.  They then create a matching
+`AuthenticationConfig` object with the `spec.type` set to `oidc` along with other
+relevant options.  This allows the API server to validate ID tokens issued by Dex.
+The user configures the `kubectl` OIDC integration and uses a Dex provided
+identity to authenticate to the cluster.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+TBD.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+#### Restrictions on asserted identities
+
+To prevent confusion with identities that are controlled by Kubernetes, the
+`system:` prefix will be disallowed in the username and groups contained in the
+`user.Info` object.  A disallowed username will cause authentication to fail.
+All disallowed groups will be filtered out.
+
+#### Avoiding ambiguity
+
+Since authorization modes such as RBAC are unaware of the origin of an identity,
+care must be taken to avoid ambiguity between identities asserted by different
+authenticators.  The `AuthenticationConfig` API will require username and groups
+prefix config to be explicitly opted out of by the cluster admin.
+
+#### Ordering of authentication modes
+
+To prevent interference with other authentication modes configured on the API
+server, all `AuthenticationConfig` objects will be considered after the other
+authenticators have had a chance to run.  This prevents issues such as a
+misbehaving authenticator attempting to authenticate a Kubernetes service
+account bearer token.  It also prevents performance issues caused by invoking
```

Comment:
Are authenticators always executed serially?

Reply:
Yes because they have a defined order they run in (i.e. SA token validation is always before a token webook).  The token cache has logic to single flight requests for the same token but this process is still serial from the perspective of a single request.